### PR TITLE
Sparse Adam with Weight Decay

### PR DESCRIPTION
**Pull request by NekuSakuraba**
Extending the idea to use sparse adam with weight decay.

- #25484 - Sparse Adam with Weight Decay

### CODE REVIEW

File: torch/optim/sparse_adam.py [hunk 1]
```diff
@@ -79,6 +79,10 @@ def make_sparse(values):
                 exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
                 beta1, beta2 = group['betas']
 
+                if group['weight_decay'] != 0:
+                    param_values = p.data.sparse_mask(grad)._values()
```
Comment (SsnL) [inline torch/optim/sparse_adam.py]: this means that you only decay the value you train, which is very very unintuitive and weird.

Comment (NekuSakuraba) [reply torch/optim/sparse_adam.py]: Thanks for your review!
Can you share an idea to improve it?

I made it like this because this optimizer is using sparse gradients and only those portions of gradients  get applied to the parameters as explained in the class docstring.

Comment (magus96) [reply torch/optim/sparse_adam.py]: Isn't this how the weight decay is implemented in Adam too?

**Comment by vincentqb** [comment]
Since Adam was originally known with L2 regularization as "weight_decay", we introduced AdamW as a separate optimizer with "new" weight decay. We should mirror this construction and introduce `sparse_adamw`.

**Comment by magus96** [comment]
Afaik a sparse neural network doesn't really need to have weight decay as the weights are already very small and there is no point in forcing them to be smaller. Do correct me if I am wrong in this analysis.
